### ISSUE TITLE
Possible issue with LZJB

### ISSUE THREAD

**Issue by fclaude** [zfsonlinux/zfs#3831]
Hi,

The decompression method lzjb_decompress has several checks ignored, and just wanted to make sure whether all of the ignores are okay in ZFSs context. I've been assisting @alkemir with some code of his for recovering some data from a ZFS partition, and we found some conditions under which the decompressor crashes.

In particular, there are some problems with decompressing, if the buffer is corrupted, in lines 119, 120 and 125. Is this code always called from a safe place where we can be sure the input won't be corrupted and that d_len is correct too?

Cheers,
Francisco.

**Comment by alkemir** [comment]
Good stuff,

pero te llevaste toda la gloria marrdito. Y ya no podremos escribir el
paper =p

Subeles el parche ;)

On Thu, Sep 24, 2015 at 12:43 PM, Francisco Claude <notifications@github.com
> wrote:

> Hi,
>
> The decompression method lzjb_decompress has several checks ignored, and
> just wanted to make sure whether all of the ignores are okay in ZFSs
> context. I've been helping @alkemir <https://github.com/alkemir> with
> some code of his for recovering some data from a ZFS partition, and I've
> had some crashes with this decompression function.
>
> In particular, there are some problems with decompressing, if the buffer
> is corrupted, in lines 119, 120 and 125. Is this code always called from a
> safe place where we can be sure the input won't be corrupted and that d_len
> is correct too?
>
> Cheers,
> Francisco.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zfsonlinux/zfs/issues/3831>.
>



-- 
Bryan Reynaert

Director
Unidad de Neurociencia Computacional
Laboratorio de Neurobiología
Universidad de Chile

**Comment by behlendorf** [comment]
@fclaude the way the code is used now by the time we get to calling this function we've already validated that the buffer has a valid checksum.  So we're as sure as we can be that the buffer isn't just garbage.  It's technically possible if quite a lot of things go wrong but it should never happen.  I'm certainly not opposed to hardening this code if we can do it without negatively impacting performance.

**Comment by alkemir** [comment]
@behlendorf  @fclaude 
Hi,

I can work on a patch proposal to harden it hopefully without a noticeable performance hit.

Cheers!

**Comment by rewolff** [comment]
I've modified the lzjb decompress as follows: 
```
/*ARGSUSED*/
int
lzjb_decompress(void *s_start, void *d_start, size_t s_len, size_t d_len, int n)
{
  uchar_t *src = s_start;
  uchar_t *dst = d_start;
  uchar_t *s_end = (uchar_t *)s_start + s_len;
  uchar_t *d_end = (uchar_t *)d_start + d_len;
  uchar_t *cpy;
  uchar_t copymap = 0;
  int copymask = 1 << (NBBY - 1);

  while ((dst < d_end) && (src < s_end)) {
    if ((copymask <<= 1) == (1 << NBBY)) {
      copymask = 1;
      copymap = *src++;
    }
```
Two  changed lines, both have s_end in them. 

I don't fully agree with your argument that "**the buffer must be ok, the checksum has been verified**". With for example a memory error occurring in the WRITING system between an update to a compressed buffer and generating the checksum, the reading system may get an invalid compressed buffer. 
On the other hand, ZFS's basic assumption is that such things don't happen: The on-disk-format is ALWAYS consistent. 

Anyway, I've added a check for input buffer overrun. I'm not sure if fclaude had other checks in mind. 
(I'm working on a data-recovery, I need to "search-for-stuff" in the raw data, so I read-a-block, decompress and search-the-result. With the output buffer being plenty-large, I was seeing enormous, implausible results)
